src/lib.rs

```diff
@@ -0,0 +1,459 @@
+    uid: RegistryUid,
+    token_id: Hash,
+    token_owner: Account,
+    metadata: Vec<u8>,
+    proof_leaves: Vec<H256>,
+}
+
+type RegistryUid = u64;
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+pub trait Trait: system::Trait + contracts::Trait + erc721::Trait + anchor::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+    /// Something that provides randomness in the runtime.
+    type Randomness: Randomness<Self::Hash>;
+
+    /// The amount of balance that must be deposited for metadata stored once.
+    type NFTDepositBase: Get<BalanceOf<Self>>;
+
+    /// The amount of balance that must be deposited per byte of metadata stored.
+    type NFTDepositPerByte: Get<BalanceOf<Self>>;
+
+    /// The amount of balance that must be deposited per validation function registry.
+    type NFTValidationRegistryDeposit: Get<BalanceOf<Self>>;
+
+    /// Currency type for this module.
+    type Currency: ReservableCurrency<Self::AccountId>
+        + LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Validation function not registered
+        ValidationFnNotExisted,
+
+        /// Uid not registered
+        UidNotRegistered,
+
+        /// Token not minted
+        TokenNotMinted,
+
+        /// Token already existed
+        TokenAlreadyExisted,
+
+        // Now token owner
+        WrongTokenOwner,
+
+        // Metadata length invalid
+        MetadataLengthInvalid,
+
+        // Not validation function
+        NotValidationFunction,
+
+        // Proof validation failed
+        ProofValidationFailure,
+
+        // Proof validation failed
+        DocumentNotAnchored,
+
+        // Token URI length
+        // TokenURILengthInvalid,
+    }
+}
+
+decl_event!(
+    pub enum Event<T> 
+        where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash {
+        // Account register a new Uid with smart contract
+        NewRegistry(AccountId, RegistryUid),
+        
+        // New NFT created
+        MintNft(RegistryUid, Hash),
+    }
+);
+
+decl_storage! {
+    trait Store for Module<T: Trait> as NftRegistry {
+        // Each registry uid include a validation function in smart contract
+        pub ValidationFn get(validator_of): map hasher(blake2_256) RegistryUid => Option<T::AccountId>;
+
+        // Next Registry id
+        pub NextRegistryId: RegistryUid;
+
+        // Registry Metadata
+        pub TokenByRegistryId get(token_by_registry_id): map T::Hash => RegistryUid;
+
+        // Metadata for each token id
+        pub RegistryTokenMetadata get(registry_token_metadata): map T::Hash => Vec<u8>;
+
+        // Metadata min length
+        pub MinMetadataLength get(min_token_metadata_length) config(): u32;
+
+        // Metadata max length
+        pub MaxMetadataLength get(max_token_metadata_length) config(): u32;
```

Comment:
@juniuszhou did you see this comment?

Reply:
Sorry, missed this one. I will put it in module's trait. I think it is better do basic check before the metadata go to validation function, because it will be passed into contract.